File: 모범 코드/9일차/9. [필수] DFS와 BFS.py
from collections import deque

def dfs(v):
    global edge, chk
    
    if(chk[v]): return
    chk[v] = True
    
    print(v, end=' ')
    
    for u in edge[v]:
        dfs(u)

N, M, V = map(int, input().split())

edge = [[] for _ in range(N+1)]
for i in range(M):
    u, v = map(int, input().split())
    edge[u].append(v)
    edge[v].append(u)

for i in range(N+1):
    edge[i].sort()

chk = [False] * (N+1)
dfs(V)
print()

chk = [False] * (N+1)
que = deque()
que.append(V)
while que:
    v = que.popleft()
    
    if(chk[v]): continue
    chk[v] = True
    print(v, end=" ")
    
    for u in edge[v]:
        que.append(u)

File: 모범 코드/9일차/9. [필수] DFS와 BFS.java
import java.io.*;
import java.util.*;

public class main{
	static boolean[] chk;
	static ArrayList<ArrayList<Integer>> edge;
	
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		
		st = new StringTokenizer(br.readLine());
		int N = Integer.parseInt(st.nextToken());
		int M = Integer.parseInt(st.nextToken());
		int V = Integer.parseInt(st.nextToken());
		
		edge = new ArrayList<ArrayList<Integer>>();
		for(int i=0; i<=N; i++) edge.add(new ArrayList<Integer>());
		while(M-- > 0){
			st = new StringTokenizer(br.readLine());
			int u = Integer.parseInt(st.nextToken());
			int v = Integer.parseInt(st.nextToken());
			
			edge.get(u).add(v);
			edge.get(v).add(u);
		}
		
		for(int i=1; i<=N; i++){
			Collections.sort(edge.get(i));
		}
		
		// dfs
		chk = new boolean[N+1];
		dfs(V);
		System.out.println();
		
		// bfs
		for(int i=1; i<=N; i++) chk[i] = false;
		Queue<Integer> que = new LinkedList<>();
		
		que.offer(V);
		while(!que.isEmpty()){
			int v = que.poll();
			
			if(chk[v]) continue;
			chk[v] = true;
			System.out.print(v + " ");
			
			for(int u: edge.get(v)) que.offer(u);
		}
	}
	
	static void dfs(int v){
		if(chk[v]) return;
		chk[v] = true;
		System.out.print(v + " ");
		
		for(int u: edge.get(v)){
			dfs(u);
		}
	}
}